1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,99 @@
+---
+RFC: RFCnnnn
+Author: Paul Higinbotham
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine
+Comments Due: July 18, 2019
+Plan to implement: Yes
+---
+
+# PowerShell ForEach-Object -Parallel Cmdlet
+
+This RFC proposes a new parameter set for the existing ForEach-Object cmdlet to parallelize script block executions, instead of running them sequentially as it does now.  
+
+## Motivation
+
+    As a PowerShell User,
+    I can do simple fan-out concurrency with the PowerShell ForEach-Object cmdlet, without having to obtain and load a separate module, or deal with PowerShell jobs unless I want to.
+
+## Specification
+
+There will be two new parameter sets added to the existing ForeEach-Object cmdlet to support both synchronous and asynchronous operations for parallel script block execution.
+For the synchronous case, the `ForEach-Object` cmdlet will not return until all parallel executions complete.
+For the asynchronous case, the `ForEach-Object` cmdlet will immediately return a PowerShell job object that contains child jobs of each parallel execution.
+
+### Implementation details
+
+Implementation will be similar to the ThreadJob module.
+Script block execution will be run for each piped input on a separate thread and runspace.

```

Comment:
How will module loading for the new runspaces be managed?  Will the runspaces be reused?  If they will be reused, will state be reset between uses?  If so, to what degree? [Just by way of ResetRunspaceState()](https://github.com/PowerShell/PowerShell/issues/7131#issuecomment-399229457)?  Or something more elaborate?  

Will ForEach-Object attempt to open modules in parallel?  If so, is it within the scope of this RFC to resolve the contention reported in PowerShell/PowerShell#7035?  If not, how will module loading occurring in the runspaces be serialized?

Reply:
Module loading will work as normal, either explicitly loaded using 'Import-Module', or implicitly loaded via command discovery.  Runspaces and threads will not be reused in the initial implementation, due to concerns of stale state causing problems, but it is something that can be looked at later.

Module loading contention is a concern but is not something that will be addressed in this work.